src/analysis/calculus/deriv.lean

```diff
@@ -1176,3 +1206,68 @@ begin
 end
 
 end
+
+section tangent_cone
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{F : Type*} [normed_group F] [normed_space 𝕜 F]
+{f : E → F} {s : set E} {f' : E →L[𝕜] F}
+
+/-- The image of a tangent cone under the differential of a map is included in the tangent cone to
+the image. -/
+lemma has_fderiv_within_at.image_tangent_cone_subset {x : E} (h : has_fderiv_within_at f f' s x) :
+  f' '' (tangent_cone_at 𝕜 s x) ⊆ tangent_cone_at 𝕜 (f '' s) (f x) :=
+begin
+  rw image_subset_iff,
+  rintros v ⟨c, d, dtop, clim, cdlim⟩,
+  refine ⟨c, (λn, f (x + d n) - f x), _, clim, _⟩,
+  show {n : ℕ | f x + (f (x + d n) - f x) ∈ f '' s} ∈ at_top,
+  { apply mem_sets_of_superset dtop,
+    simp [-mem_image, mem_image_of_mem] {contextual := tt} },
+  show tendsto (λ (n : ℕ), c n • (f (x + d n) - f x)) at_top (nhds (f' v)),
+  { apply h.lim dtop clim cdlim, }
+end
```

Comment:
The above proof is confusing to read. the refine spawns two goals but you don't use blocks to delimit them. The `apply` from the last line could be an exact (conveying the extra information that there is nothing implicit here). I would prefer the last two lines to read:
```lean
  show tendsto (λ (n : ℕ), c n • (f (x + d n) - f x)) at_top (nhds (f' v)),
    from h.lim dtop clim cdlim, 
```
You could also replace the whole thing by:
```lean
  rw image_subset_iff,
  rintros v ⟨c, d, dtop, clim, cdlim⟩,
  refine ⟨c, (λn, f (x + d n) - f x), _, clim, h.lim dtop clim cdlim⟩,
  apply mem_sets_of_superset dtop,
  simp [-mem_image, mem_image_of_mem] {contextual := tt}
```
but this is probably less readable.

Reply:
Compressed it even more. For such a short proof, readability is not really important.